File: Sources/Automerge/TextEncoding.swift
import enum AutomergeUniffi.TextEncoding
import Foundation

/// An enumeration representing different types of text encoding.
///
/// Text encodings determine how text is processed across Automerge APIs.
public enum TextEncoding {
    /// Text encoding using Grapheme Cluster.
    /// Grapheme clusters represent user-perceived characters, which may consist of multiple Unicode scalars.
    /// For example:
    /// - "é" (a Latin small letter "e" followed by a combining acute accent) is a single grapheme cluster.
    /// - Emoji with modifiers, like "👨‍👩‍👧‍👦" (family emoji), are grapheme clusters combining multiple scalars.
    case graphemeCluster

    /// Text encoding using Unicode scalar values.
    /// Unicode scalars are the fundamental building blocks of Unicode text, but they are not directly stored as-is.
    /// Instead, they are encoded into binary formats like UTF-8 or UTF-16 for persistence.
    case unicodeScalar

    /// Text encoding using UTF-8.
    /// A variable-width encoding representing characters in 1–4 bytes.
    case utf8

    /// Text encoding using UTF-16.
    /// A variable-width encoding using one or two 16-bit code units.
    case utf16
}

// MARK: - Adapters
typealias FfiTextEncoding = AutomergeUniffi.TextEncoding

extension FfiTextEncoding {
    var textEncoding: TextEncoding {
        switch self {
        case .graphemeCluster: return .graphemeCluster
        case .unicodeCodePoint: return .unicodeScalar
        case .utf16CodeUnit: return .utf16
        case .utf8CodeUnit: return .utf8
        }
    }
}

extension TextEncoding {
    var ffi_textEncoding: FfiTextEncoding {
        switch self {
        case .graphemeCluster: return .graphemeCluster
        case .unicodeScalar: return .unicodeCodePoint
        case .utf16: return .utf16CodeUnit
        case .utf8: return .utf8CodeUnit
        }
    }
}

File: rust/src/text_encoding.rs
use automerge as am;

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TextEncoding {
    UnicodeCodePoint,
    Utf8CodeUnit,
    Utf16CodeUnit,
    GraphemeCluster,
}

impl From<TextEncoding> for am::TextEncoding {
    fn from(value: TextEncoding) -> Self {
        match value {
            TextEncoding::UnicodeCodePoint => am::TextEncoding::UnicodeCodePoint,
            TextEncoding::Utf8CodeUnit => am::TextEncoding::Utf8CodeUnit,
            TextEncoding::Utf16CodeUnit => am::TextEncoding::Utf16CodeUnit,
            TextEncoding::GraphemeCluster => am::TextEncoding::GraphemeCluster,
        }
    }
}

impl From<am::TextEncoding> for TextEncoding {
    fn from(value: am::TextEncoding) -> Self {
        match value {
            am::TextEncoding::UnicodeCodePoint => TextEncoding::UnicodeCodePoint,
            am::TextEncoding::Utf8CodeUnit => TextEncoding::Utf8CodeUnit,
            am::TextEncoding::Utf16CodeUnit => TextEncoding::Utf16CodeUnit,
            am::TextEncoding::GraphemeCluster => TextEncoding::GraphemeCluster,
        }
    }
}

impl std::fmt::Display for TextEncoding {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnicodeCodePoint => write!(f, "Unicode Scalar"),
            Self::Utf8CodeUnit => write!(f, "UTF-8"),
            Self::Utf16CodeUnit => write!(f, "UTF-16"),
            Self::GraphemeCluster => write!(f, "Grapheme Cluster"),
        }
    }
}
